warp/src/raygun/community.rs

```diff
@@ -0,0 +1,224 @@
+        &self.name
+    }
+    pub fn description(&self) -> &Option<String> {
+        &self.description
+    }
+    pub fn created(&self) -> DateTime<Utc> {
+        self.created
+    }
+    pub fn modified(&self) -> DateTime<Utc> {
+        self.modified
+    }
+    pub fn channel_type(&self) -> CommunityChannelType {
+        self.channel_type
+    }
+    pub fn permissions(&self) -> &CommunityChannelPermissions {
+        &self.permissions
+    }
+}
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+pub enum CommunityChannelType {
+    Text,
+    Voice,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum CommunityPermission {
+    EditName,
+    EditDescription,
+    ManageRoles,
+    ManagePermissions,
+    ManageMembers,
+    ManageChannels,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum CommunityChannelPermission {
+    ViewChannel,
+    EditName,
+    EditDescription,
+    SendMessages,
+    DeleteMessages,
+}
+
+#[async_trait::async_trait]
+pub trait RayGunCommunity: Sync + Send {
+    async fn create_community(&mut self, _owner: &DID, _name: &str) -> Result<Community, Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn delete_community(&mut self, _community_id: Uuid) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn get_community(&mut self, _community_id: Uuid) -> Result<Community, Error> {
+        Err(Error::Unimplemented)
+    }
+
+    async fn create_channel(
+        &mut self,
+        _community_id: Uuid,
+        _channel_name: &str,
+        _channel_type: CommunityChannelType,
+    ) -> Result<CommunityChannel, Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn delete_channel(&mut self, _community_id: Uuid, _channel_id: Uuid) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn get_channel(
+        &mut self,
+        _community_id: Uuid,
+        _channel_id: Uuid,
+    ) -> Result<CommunityChannel, Error> {
+        Err(Error::Unimplemented)
+    }
+
+    async fn edit_community_name(&mut self, _community_id: Uuid, _name: &str) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn edit_community_description(
+        &mut self,
+        _community_id: Uuid,
+        _description: Option<String>,
+    ) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn edit_community_roles(
+        &mut self,
+        _community_id: Uuid,
+        _roles: IndexSet<Role>,
+    ) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn edit_community_permissions(
+        &mut self,
+        _community_id: Uuid,
+        _permissions: CommunityPermissions,
+    ) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
+    async fn add_community_member(&mut self, _community_id: Uuid, _member: DID) -> Result<(), Error> {
+        Err(Error::Unimplemented)
+    }
```

Comment:
I wonder if this should have a option to treat it as an invite of sorts so the user would have the option to join or ignore. Thoughts?